File: android/app/src/test/java/org/vectair/vcas/car/logic/RelevanceTest.kt
package org.vectair.vcas.car.logic

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import kotlin.math.cos

/**
 * Real JUnit4 verification of the relevance.js -> Relevance.kt port,
 * following this project's established discipline (see CLAUDE.md,
 * "Android Auto phase 1, continued" and the two logic ports before this
 * one): run against real execution, not just read for correctness.
 *
 * Three verification styles:
 *  - Exact-boundary tests for the teardrop shape itself (0/60/180 deg),
 *    using values solved analytically from the same closed-form formula
 *    teardropRangeNm() uses, so the `<=` cutoff can be checked precisely
 *    without any geodesic computation being involved at all — evaluate()
 *    accepts relativeBearing/vis.slantRangeNm directly, so these need no
 *    real coordinates.
 *  - Scenario tests (overhead, contrail range extension, "extension never
 *    shrinks a larger custom rMaxNm") against the real documented branch
 *    behaviour, each worked through by hand before asserting.
 *  - Predicted-entry tests, which DO need real coordinates since
 *    predictedEntrySeconds() calls Geo internally: rather than trusting
 *    hand-picked margins, a small helper replicates the same
 *    project/bearing/range computation using Geo's own already-verified
 *    (GeoTest.kt, 34/34 passing) primitives to independently determine
 *    which sample should trigger, then checks Relevance.evaluate()
 *    against that — not a re-test of Geo, a check that
 *    predictedEntrySeconds()'s own sampling loop (step size, sample
 *    count, early-return) matches the documented algorithm.
 */
class RelevanceTest {

    private val userLat = 40.0
    private val userLon = -75.0

    // ---- Exact teardrop-boundary tests (via evaluate(), no real coords) ----

    private fun visAt(slantRangeNm: Double, elevationDeg: Double = 0.0) = Relevance.VisInput(slantRangeNm, elevationDeg)

    // An aircraft with no track/speed data can never trigger predicted-entry,
    // so any evaluate() call using this guarantees only the in-view/overhead
    // branches are exercised.
    private fun acNoMotionData(altitudeFt: Double? = null) =
        Relevance.AircraftState(lat = userLat, lon = userLon, altitudeFt = altitudeFt)

    private val stationaryUser = Relevance.UserState(userLat, userLon, heading = 0.0, speedMph = 0.0)

    @Test
    fun teardrop_deadAhead_justInsideBoundary() {
        // teardropRangeNm(0) == rMaxNm == 15.0 exactly (cos(0)=1, f=1).
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 0.0, visAt(15.0))
        assertEquals(true, result.relevant)
        assertEquals("in-view", result.reason)
        assertEquals(0.0, result.enterInSeconds!!, 1e-9)
    }

    @Test
    fun teardrop_deadAhead_justOutsideBoundary() {
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 0.0, visAt(15.01))
        assertEquals(false, result.relevant)
        assertNull(result.reason)
        assertNull(result.enterInSeconds)
    }

    @Test
    fun teardrop_deadBehind_justInsideBoundary() {
        // teardropRangeNm(180) == rMinNm == 3.0 exactly (cos(180)=-1, f=0).
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 180.0, visAt(3.0))
        assertEquals(true, result.relevant)
        assertEquals("in-view", result.reason)
    }

    @Test
    fun teardrop_deadBehind_justOutsideBoundary() {
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 180.0, visAt(3.01))
        assertEquals(false, result.relevant)
    }

    @Test
    fun teardrop_at60Degrees_exactAnalyticBoundary() {
        // cos(60deg)=0.5, f=((1+0.5)/2)^2=0.5625, range = 3 + 12*0.5625 = 9.75 exactly.
        val boundary = Relevance.DEFAULTS.rMinNm +
            (Relevance.DEFAULTS.rMaxNm - Relevance.DEFAULTS.rMinNm) *
            ((1 + cos(Math.toRadians(60.0))) / 2).let { Math.pow(it, Relevance.DEFAULTS.pinchExponent) }
        assertEquals(9.75, boundary, 1e-9)

        val inside = Relevance.evaluate(stationaryUser, acNoMotionData(), 60.0, visAt(boundary))
        assertEquals("in-view", inside.reason)

        val outside = Relevance.evaluate(stationaryUser, acNoMotionData(), 60.0, visAt(boundary + 0.01))
        assertEquals(false, outside.relevant)
    }

    // ---- Overhead ----

    @Test
    fun overhead_justAboveThreshold_isRelevantRegardlessOfBearingOrRange() {
        // elevationDeg > 70, with bearing/range that would otherwise be
        // nowhere near the teardrop.
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 90.0, visAt(slantRangeNm = 999.0, elevationDeg = 70.01))
        assertEquals(true, result.relevant)
        assertEquals("overhead", result.reason)
        assertEquals(0.0, result.enterInSeconds!!, 1e-9)
    }

    @Test
    fun overhead_exactlyAtThreshold_doesNotQualify() {
        // Condition is strictly `>`, not `>=`.
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(), 90.0, visAt(slantRangeNm = 999.0, elevationDeg = 70.0))
        assertEquals(false, result.relevant)
    }

    // ---- Contrail range extension ----

    @Test
    fun highAltitude_extendsDeadAheadRangeTo50nm() {
        // 40nm is beyond the base 15nm teardrop but within the 50nm
        // contrail extension cap, for an aircraft at/above 26,000ft.
        val result = Relevance.evaluate(
            stationaryUser,
            acNoMotionData(altitudeFt = 30000.0),
            0.0,
            visAt(40.0)
        )
        assertEquals(true, result.relevant)
        assertEquals("in-view", result.reason)
    }

    @Test
    fun belowContrailAltitude_doesNotExtendRange() {
        val result = Relevance.evaluate(
            stationaryUser,
            acNoMotionData(altitudeFt = 10000.0),
            0.0,
            visAt(40.0)
        )
        assertEquals(false, result.relevant)
    }

    @Test
    fun contrailExtension_neverShrinksALargerCustomRMaxNm() {
        // A custom rMaxNm (60) bigger than the default rangeExtensionCapNm
        // (50) must not be clamped DOWN to 50 just because the contrail
        // condition also applies — effectiveRMaxNm is max(rMaxNm, cap).
        val customOpts = Relevance.Options(rMaxNm = 60.0)
        val result = Relevance.evaluate(
            stationaryUser,
            acNoMotionData(altitudeFt = 30000.0),
            0.0,
            visAt(55.0),
            customOpts
        )
        assertEquals(true, result.relevant)
        assertEquals("in-view", result.reason)
    }

    @Test
    fun altitudeNull_doesNotExtendRange() {
        val result = Relevance.evaluate(stationaryUser, acNoMotionData(altitudeFt = null), 0.0, visAt(40.0))
        assertEquals(false, result.relevant)
    }

    // ---- Predicted entry: insufficient data ----

    @Test
    fun predictedEntry_missingTrackDeg_neverTriggers() {
        val aircraft = Relevance.AircraftState(userLat, userLon, trackDeg = null, groundSpeedKt = 250.0)
        val result = Relevance.evaluate(stationaryUser, aircraft, 0.0, visAt(999.0))
        assertEquals(false, result.relevant)
    }

    @Test
    fun predictedEntry_missingGroundSpeed_neverTriggers() {
        val aircraft = Relevance.AircraftState(userLat, userLon, trackDeg = 180.0, groundSpeedKt = null)
        val result = Relevance.evaluate(stationaryUser, aircraft, 0.0, visAt(999.0))
        assertEquals(false, result.relevant)
    }

    // ---- Predicted entry: real convergence, verified against an
    // independent replica built from Geo's own verified primitives ----

    /**
     * Replicates predictedEntrySeconds()'s own per-sample projection
     * exactly (same formulas, freshly written here rather than copied)
     * so the expected trigger sample can be derived independently of
     * Relevance.kt's own implementation, using Geo functions already
     * proven correct by GeoTest.kt.
     */
    private fun independentEntrySeconds(
        userState: Relevance.UserState,
        aircraft: Relevance.AircraftState,
        opts: Relevance.Options
    ): Double? {
        val acSpeedMps = aircraft.groundSpeedKt!! * 0.514444
        val userSpeedMps = (userState.speedMph ?: 0.0) * 0.44704
        val userIsMoving = (userState.speedMph ?: 0.0) >= opts.stationarySpeedMph
        val stepSeconds = opts.lookaheadSeconds / opts.lookaheadSamples

        for (i in 1..opts.lookaheadSamples) {
            val t = stepSeconds * i
            val acPos = Geo.projectPosition(aircraft.lat, aircraft.lon, aircraft.trackDeg!!, acSpeedMps * t)
            val userPos = if (userIsMoving) {
                Geo.projectPosition(userState.lat, userState.lon, userState.heading, userSpeedMps * t)
            } else {
                Geo.LatLon(userState.lat, userState.lon)
            }
            val bearing = Geo.calculateBearing(userPos.lat, userPos.lon, acPos.lat, acPos.lon)
            val relBearing = Geo.calculateRelativeBearing(bearing, userState.heading)
            val rangeNm = Geo.calculateDistanceNm(userPos.lat, userPos.lon, acPos.lat, acPos.lon)

            val c = cos(Math.toRadians(relBearing))
            val f = Math.pow((1 + c) / 2, opts.pinchExponent)
            val boundary = opts.rMinNm + (opts.rMaxNm - opts.rMinNm) * f
            if (rangeNm <= boundary) return t
        }
        return null
    }

    @Test
    fun predictedEntry_aircraftConvergingOnStationaryUser_matchesIndependentReplica() {
        // Aircraft 22nm dead ahead (north) of a stationary user, flying
        // south (track 180) directly toward the user at a speed high
        // enough to close inside the 15nm dead-ahead teardrop within the
        // 15s lookahead window.
        val acStart = Geo.destinationPoint(userLat, userLon, 0.0, 22.0 * 1852.0)
        val aircraft = Relevance.AircraftState(acStart.lat, acStart.lon, trackDeg = 180.0, groundSpeedKt = 1800.0)

        val currentBearing = Geo.calculateBearing(userLat, userLon, acStart.lat, acStart.lon)
        val currentRelBearing = Geo.calculateRelativeBearing(currentBearing, stationaryUser.heading)
        val currentRangeNm = Geo.calculateDistanceNm(userLat, userLon, acStart.lat, acStart.lon)

        val expected = independentEntrySeconds(stationaryUser, aircraft, Relevance.DEFAULTS)
        assertNotNull("test setup must actually converge within the lookahead window", expected)

        val result = Relevance.evaluate(stationaryUser, aircraft, currentRelBearing, visAt(currentRangeNm))
        assertEquals(true, result.relevant)
        assertEquals("predicted-entry", result.reason)
        assertEquals(expected!!, result.enterInSeconds!!, 1e-9)
    }

    @Test
    fun predictedEntry_movingUserApproachingStationaryAircraft_matchesIndependentReplica() {
        // Stationary aircraft 15.5nm dead ahead (north); the USER moves
        // toward it (exercising the userIsMoving=true projection branch,
        // not covered by the stationary-user test above).
        val movingUser = Relevance.UserState(userLat, userLon, heading = 0.0, speedMph = 150.0)
        val acPos = Geo.destinationPoint(userLat, userLon, 0.0, 15.5 * 1852.0)
        val aircraft = Relevance.AircraftState(acPos.lat, acPos.lon, trackDeg = 0.0, groundSpeedKt = 0.0)

        val currentBearing = Geo.calculateBearing(userLat, userLon, acPos.lat, acPos.lon)
        val currentRelBearing = Geo.calculateRelativeBearing(currentBearing, movingUser.heading)
        val currentRangeNm = Geo.calculateDistanceNm(userLat, userLon, acPos.lat, acPos.lon)

        val expected = independentEntrySeconds(movingUser, aircraft, Relevance.DEFAULTS)
        assertNotNull("test setup must actually converge within the lookahead window", expected)

        val result = Relevance.evaluate(movingUser, aircraft, currentRelBearing, visAt(currentRangeNm))
        assertEquals(true, result.relevant)
        assertEquals("predicted-entry", result.reason)
        assertEquals(expected!!, result.enterInSeconds!!, 1e-9)
    }

    @Test
    fun predictedEntry_aircraftDivergingFromUser_neverTriggers() {
        // Same setup as the first convergence test, but the aircraft
        // flies AWAY (track 0, continuing north) instead of toward the
        // user — range only increases, so no sample should ever qualify.
        val acStart = Geo.destinationPoint(userLat, userLon, 0.0, 22.0 * 1852.0)
        val aircraft = Relevance.AircraftState(acStart.lat, acStart.lon, trackDeg = 0.0, groundSpeedKt = 1800.0)

        val currentBearing = Geo.calculateBearing(userLat, userLon, acStart.lat, acStart.lon)
        val currentRelBearing = Geo.calculateRelativeBearing(currentBearing, stationaryUser.heading)
        val currentRangeNm = Geo.calculateDistanceNm(userLat, userLon, acStart.lat, acStart.lon)

        val expected = independentEntrySeconds(stationaryUser, aircraft, Relevance.DEFAULTS)
        assertNull("diverging aircraft must not converge in the independent replica either", expected)

        val result = Relevance.evaluate(stationaryUser, aircraft, currentRelBearing, visAt(currentRangeNm))
        assertEquals(false, result.relevant)
        assertNull(result.reason)
        assertNull(result.enterInSeconds)
    }

    // ---- DEFAULTS ----

    @Test
    fun defaults_matchDocumentedTunedConstants() {
        assertEquals(15.0, Relevance.DEFAULTS.rMaxNm, 1e-9)
        assertEquals(3.0, Relevance.DEFAULTS.rMinNm, 1e-9)
        assertEquals(2.0, Relevance.DEFAULTS.pinchExponent, 1e-9)
        assertEquals(70.0, Relevance.DEFAULTS.overheadElevationDeg, 1e-9)
        assertEquals(15.0, Relevance.DEFAULTS.lookaheadSeconds, 1e-9)
        assertEquals(3, Relevance.DEFAULTS.lookaheadSamples)
        assertEquals(5.0, Relevance.DEFAULTS.stationarySpeedMph, 1e-9)
        assertEquals(26000.0, Relevance.DEFAULTS.contrailMinAltitudeFt, 1e-9)
        assertEquals(50.0, Relevance.DEFAULTS.rangeExtensionCapNm, 1e-9)
    }
}

File: tests/logic/relevance.test.js
/**
 * Real-execution checks for src/logic/relevance.js — the teardrop
 * relevance filter deciding which aircraft are worth showing at all,
 * independent of how visible they'd be once looked at. See CLAUDE.md's
 * "Contrail visibility — Relevance range + Visibility score" entry for the
 * real field-confirmed gap (a fixed 15nm cap excluding a genuinely
 * identifiable high-altitude jet) that motivated the high-altitude range
 * extension tested below.
 */
const { createSuite } = require("../support/assert");
const { loadLogic } = require("../support/loadLogic");

const { Geo, Relevance } = loadLogic();
const t = createSuite("relevance.js");

const BASE_USER_STATE = { lat: 0, lon: 0, heading: 0, speedMph: 0 };

function vis(slantRangeNm, elevationDeg) {
  return { slantRangeNm, elevationDeg };
}

function aircraft(overrides) {
  return Object.assign({ lat: 0, lon: 0, altitudeFt: 5000, trackDeg: null, groundSpeedKt: null }, overrides);
}

// ---- overhead override ----
{
  const r = Relevance.evaluate(BASE_USER_STATE, aircraft(), 170, vis(999, 75));
  t.eq(r.relevant, true, "elevation > 70deg is relevant regardless of bearing/range");
  t.eq(r.reason, "overhead", "the overhead reason is reported");
  t.eq(r.enterInSeconds, 0, "overhead is immediate (enterInSeconds 0)");

  const rBoundary = Relevance.evaluate(BASE_USER_STATE, aircraft(), 0, vis(3, 70));
  t.ok(rBoundary.reason !== "overhead", "elevation exactly 70deg is NOT overhead (strictly '> 70', not '>= 70')");
}

// ---- teardrop shape: exact closed-form values at default opts ----
{
  // Dead ahead (0deg): boundary is exactly rMaxNm (15).
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 0, vis(15, 0)).relevant, true,
    "dead-ahead range exactly at rMaxNm (15nm) is relevant (inclusive boundary)");
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 0, vis(15.5, 0)).relevant, false,
    "dead-ahead range just beyond rMaxNm is not relevant (no track data, so no predicted-entry rescue either)");

  // Dead behind (180deg): boundary is exactly rMinNm (3).
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 180, vis(3, 0)).relevant, true,
    "dead-behind range exactly at rMinNm (3nm) is relevant");
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 180, vis(3.5, 0)).relevant, false,
    "dead-behind range just beyond rMinNm is not relevant");

  // 60deg, pinchExponent=2: documented exact value from CLAUDE.md's own
  // Kotlin-port test suite writeup — teardropRangeNm(60) == 9.75 exactly.
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 60, vis(9.75, 0)).relevant, true,
    "60deg boundary matches the documented exact teardrop value (9.75nm)");
  t.eq(Relevance.evaluate(BASE_USER_STATE, aircraft(), 60, vis(9.76, 0)).relevant, false,
    "just beyond the 60deg boundary is not relevant");
}

// ---- high-altitude range extension ----
{
  const highAlt = aircraft({ altitudeFt: 30000 }); // >= contrailMinAltitudeFt (26000)
  const lowAlt = aircraft({ altitudeFt: 5000 });    // below the threshold

  const rExtended = Relevance.evaluate(BASE_USER_STATE, highAlt, 0, vis(40, 0));
  t.eq(rExtended.relevant, true, "a 30000ft aircraft at 40nm dead-ahead IS relevant (extended teardrop, cap 50)");

  const rNotExtended = Relevance.evaluate(BASE_USER_STATE, lowAlt, 0, vis(40, 0));
  t.eq(rNotExtended.relevant, false,
    "the SAME 40nm dead-ahead range is NOT relevant at 5000ft (base rMaxNm=15, no extension) — proves the extension is genuinely conditional on altitude");

  // Altitude eligibility boundary: exactly at vs just below contrailMinAltitudeFt.
  const rAtThreshold = Relevance.evaluate(BASE_USER_STATE, aircraft({ altitudeFt: 26000 }), 0, vis(40, 0));
  const rBelowThreshold = Relevance.evaluate(BASE_USER_STATE, aircraft({ altitudeFt: 25999 }), 0, vis(40, 0));
  t.eq(rAtThreshold.relevant, true, "altitude exactly at the 26000ft threshold IS extended (inclusive '>=')");
  t.eq(rBelowThreshold.relevant, false, "altitude 1ft below the threshold is NOT extended");

  // A custom rMaxNm larger than the extension cap survives unshrunk.
  const rCustomCap = Relevance.evaluate(BASE_USER_STATE, highAlt, 0, vis(58, 0), { rMaxNm: 60 });
  t.eq(rCustomCap.relevant, true,
    "a custom rMaxNm (60) larger than rangeExtensionCapNm (50) is not clamped down to 50 (max(rMaxNm, cap))");
}

// ---- predicted-entry: cross-checked against an independent reimplementation ----
// Deliberately NOT copied from relevance.js's own _predictedEntrySeconds —
// freshly written against the documented algorithm using only Geo's own
// already-verified public functions, so this proves the real
// implementation's specific sampling behaviour (step size, sample count,
// early-return-on-first-match) against independently-derived ground truth,
// not against itself. Mirrors this project's own established pattern for
// this exact function (see CLAUDE.md's Kotlin RelevanceTest writeup).
function independentEntrySeconds(userState, ac, opts) {
  const stepSeconds = opts.lookaheadSeconds / opts.lookaheadSamples;
  const userIsMoving = (userState.speedMph || 0) >= opts.stationarySpeedMph;
  for (let i = 1; i <= opts.lookaheadSamples; i++) {
    const tSec = stepSeconds * i;
    const acSpeedMps = ac.groundSpeedKt * 0.514444;
    const acPos = Geo.projectPosition(ac.lat, ac.lon, ac.trackDeg, acSpeedMps * tSec);
    const userPos = userIsMoving
      ? Geo.projectPosition(userState.lat, userState.lon, userState.heading, (userState.speedMph || 0) * 0.44704 * tSec)
      : { lat: userState.lat, lon: userState.lon };
    const bearing = Geo.calculateBearing(userPos.lat, userPos.lon, acPos.lat, acPos.lon);
    const relBearing = Geo.calculateRelativeBearing(bearing, userState.heading);
    const rangeNm = Geo.calculateDistanceNm(userPos.lat, userPos.lon, acPos.lat, acPos.lon);
    const rad = (relBearing * Math.PI) / 180;
    const f = Math.pow((1 + Math.cos(rad)) / 2, opts.pinchExponent);
    const teardropRangeNm = opts.rMinNm + (opts.rMaxNm - opts.rMinNm) * f;
    if (rangeNm <= teardropRangeNm) return tSec;
  }
  return null;
}

{
  const opts = Relevance.DEFAULTS;

  // A converging, stationary-observer scenario: aircraft starts just
  // outside the teardrop, heading toward the observer.
  const start = Geo.destinationPoint(0, 0, 0, 16 * 1852); // 16nm dead north, just beyond rMaxNm=15
  const convergingAc = aircraft({ lat: start.lat, lon: start.lon, trackDeg: 180, groundSpeedKt: 300 });
  const rConverging = Relevance.evaluate(BASE_USER_STATE, convergingAc, 0, vis(16, 0), opts);
  const expectedConverging = independentEntrySeconds(BASE_USER_STATE, convergingAc, opts);
  t.ok(expectedConverging !== null, "sanity check: the independent reimplementation itself finds a convergence sample");
  t.eq(rConverging.relevant, true, "a converging aircraft is correctly found relevant via predicted-entry");
  t.eq(rConverging.reason, "predicted-entry", "the predicted-entry reason is reported");
  t.eq(rConverging.enterInSeconds, expectedConverging,
    "enterInSeconds exactly matches an independent reimplementation of the same sampling algorithm");

  // A moving-user scenario, exercising the userIsMoving branch.
  const movingUserState = { lat: 0, lon: 0, heading: 0, speedMph: 40 };
  const rConvergingMoving = Relevance.evaluate(movingUserState, convergingAc, 0, vis(16, 0), opts);
  const expectedMoving = independentEntrySeconds(movingUserState, convergingAc, opts);
  t.eq(rConvergingMoving.enterInSeconds, expectedMoving,
    "the userIsMoving branch also exactly matches the independent reimplementation");

  // A diverging aircraft (flying away) never converges within the window.
  const divergingAc = aircraft({ lat: start.lat, lon: start.lon, trackDeg: 0, groundSpeedKt: 300 });
  const rDiverging = Relevance.evaluate(BASE_USER_STATE, divergingAc, 0, vis(16, 0), opts);
  t.eq(rDiverging.relevant, false, "a diverging aircraft is correctly found NOT relevant");
  t.eq(rDiverging.reason, null, "a not-relevant result carries a null reason");
  t.eq(rDiverging.enterInSeconds, null, "a not-relevant result carries a null enterInSeconds");

  // Missing track/speed data short-circuits predicted-entry to null.
  const noTrackAc = aircraft({ lat: start.lat, lon: start.lon, trackDeg: null, groundSpeedKt: 300 });
  t.eq(Relevance.evaluate(BASE_USER_STATE, noTrackAc, 0, vis(16, 0), opts).relevant, false,
    "missing trackDeg short-circuits predicted-entry (never relevant via that path)");
  const noSpeedAc = aircraft({ lat: start.lat, lon: start.lon, trackDeg: 180, groundSpeedKt: null });
  t.eq(Relevance.evaluate(BASE_USER_STATE, noSpeedAc, 0, vis(16, 0), opts).relevant, false,
    "missing groundSpeedKt short-circuits predicted-entry (never relevant via that path)");
}

// ---- options override merges with DEFAULTS, doesn't wipe them ----
{
  const r = Relevance.evaluate(BASE_USER_STATE, aircraft(), 0, vis(5, 0), { rMaxNm: 5 });
  t.eq(r.relevant, true, "a custom rMaxNm (5) is honoured at dead-ahead range 5nm");
  const rBehind = Relevance.evaluate(BASE_USER_STATE, aircraft(), 180, vis(3, 0), { rMaxNm: 5 });
  t.eq(rBehind.relevant, true,
    "overriding only rMaxNm leaves rMinNm (3) at its DEFAULTS value, not wiped out by Object.assign");
}

t.done();
